clients/drcachesim/simulator/cache_lru.cpp

```diff
@@ -62,6 +62,12 @@ cache_lru_t::init(int associativity, int block_size, int total_size,
     return true;
 }
 
+void
+cache_lru_t::invalidate_update(caching_device_block_t *block)
+{
+    // Leave counters unchanged for LRU.
```

Comment:
I've removed the function now (based on your earlier comment), which makes it difficult to put a comment in.

I'm probably biased from having stared at this for a while, but I think it's fairly clear the LRU implementation very much relies on the counter values of invalid lines. Otherwise, why initialise the counters in each set to have unique values at the start (`cache_lru_t::init`), given they're all invalid lines?

Reply:
I puzzled over that for a while, too.  I believe it's because the line replacement code expects the way counters to be unique (to specify a full ordering) at all times, even for invalid lines.  So the init code sets this property and everything else has to maintain it.
The FIFO code expects the counters to maintain the property of exactly 1 non-zero counter, and again breaks if the property is violated.